zulipterminal/helper.py

```diff
@@ -459,15 +460,71 @@ def match_emoji(emoji: str, text: str) -> bool:
     return emoji.lower().startswith(text.lower())
 
 
-def match_stream(stream: Any, text: str) -> bool:
+StreamT = TypeVar('StreamT')
+
+
+def match_stream(streams: List[Tuple[StreamT, str]], text: str,
+                 pinned_streams: List[List[Any]]) -> List[StreamT]:
     """
-    True if the stream matches with `text` (case insensitive),
-    False otherwise.
+    Returns a list of streams (StreamT) whose words match with the 'text' in
+    the following order:
+    * 1st-word startswith match > 2nd-word startswith match > ... (pinned)
+    * 1st-word startswith match > 2nd-word startswith match > ... (unpinned)
+
+    Data structures:
+    * `stream_splits` is a list of tuples where each tuple T ->
+      (stream, [stream_name, second_word, third_word, ...]).
+    * `matches` is an OrderedDict which stores streams (StreamT) that match in
+      the following manner:
+      matches = {
+          'pinned': {
+              0: [streams (StreamT) whose first word matches],  # If any.
+              1: [streams (StreamT) whose second word matches], # If any.
+              ...
+          },
+          # Similarly for 'unpinned'.
+      }
+
+    Note: This function expects `streams` to be sorted in a non-decreasing
+    order.
     """
-    stream_name = stream[0].lower()
-    if stream_name.startswith(text.lower()):
-        return True
-    return False
+    pinned_stream_names = [stream[0] for stream in pinned_streams]
+
+    # Note that 'space' is put in `delimiters` only to show the types of
+    # delimiters that are supported. Otherwise, it is redundant (see the
+    # following .replace method).
+    delimiters = [' ', '-', '_', '/']
+    stream_splits = []
+    for stream, stream_name in streams:
+        delimit_with_spaces = stream_name
+        for delimiter in delimiters:
+            delimit_with_spaces = delimit_with_spaces.replace(delimiter, ' ')
+        stream_splits.append(
+            (stream, [stream_name] + delimit_with_spaces.split()[1:])
+        )
+
+    matches = OrderedDict([
+        ('pinned', defaultdict(list)),
+        ('unpinned', defaultdict(list)),
+    ])  # type: OrderedDict[str, DefaultDict[int, List[StreamT]]]
+
+    for stream, splits in stream_splits:
+        stream_name = splits[0]
+        for match_position, word in enumerate(splits):
+            if word.lower().startswith(text.lower()):
+                kind = ('pinned' if stream_name in pinned_stream_names
+                        else 'unpinned')
```

Comment:
We generate `kind` here, but it only depends on the `stream_name` at the top of the first loop. Perhaps you were wanting to avoid calculating `kind` until it was definitely used? If we move it up it should make it clearer that kind doesn't change depending upon what's in the lower loop/condition.

The pinned logic could be taken further and we could pre-sort the pinned and unpinned, but let's skip that for now, unless you find it looks even tidier!

Reply:
Indeed, I did that to avoid calculating `kind` every time. However, the suggested approach is much cleaner. :+1: